src/transformers/generation/utils.py

```diff
@@ -4905,3 +4914,124 @@ def _ranking_fast(
     contrastive_score = torch.stack(torch.split(contrastive_score, beam_width))  # [B, K]
     _, selected_idx = contrastive_score.max(dim=-1)  # [B]
     return selected_idx
+
+
+def _split_model_inputs(
+    model_output: Union[ModelOutput, Dict], split_size: int, full_batch_size: int
+) -> List[Union[ModelOutput, Dict]]:
+    """
+    Split a ModelOutput object (or its subclasses) or Dict into a list of same-class objects based on a specified split
+    size. The input object is dict when it was prepared for forward pass and ModelOutput when it was returned from
+    previous forward pass.
+    """
+    # Infer the class from the object
+    model_output_cls = type(model_output)
+    if (full_batch_size % split_size) != 0:
+        raise ValueError("`full_batch_size` must be divisible by `split_size`")
+
+    if split_size > full_batch_size:
+        raise ValueError("`split_size` must be smaller or equal to `full_batch_size`")
+
+    # Helper function to split tensors or tuples of tensors
+    def _split(data):
+        """
+        Takes care of three cases:
+        1. data is a tensor: e.g. last_hidden_state, pooler_output etc. split them on the batch_size dim
+        2. data is a tuple: e.g. hidden_states, attentions etc. Keep the tuple as it is and split each tensor in it and
+           return a list of tuples
+        3. data is a tuple of tuples, e.g. past_key_values. Keep the tuple as it is and split each tuple in it and
+           return a list of tuples of tuples
+        (see documentation of ModelOutput)
+        """
+        if data is None:
+            return [None] * (full_batch_size // split_size)
+        if isinstance(data, torch.Tensor):
+            return [data[i : i + split_size] for i in range(0, full_batch_size, split_size)]
+        elif isinstance(data, tuple):
+            # If the elements of the tuple are also tuples (e.g., past_key_values in our earlier example)
+            if isinstance(data[0], tuple):
+                return [
+                    tuple(tuple(tensor[i : i + split_size] for tensor in inner_tuple) for inner_tuple in data)
+                    for i in range(0, full_batch_size, split_size)
+                ]
+
+            else:
+                return [
+                    tuple(sub_tensor[i : i + split_size] for sub_tensor in data)
+                    for i in range(0, full_batch_size, split_size)
+                ]
+        elif isinstance(data, ModelOutput):
+            return _split_model_inputs(model_output["encoder_outputs"], split_size, full_batch_size)
```

Comment:
bit strange to define and call in the same function let's maybe define `_split` outside

Reply:
I am also not a super fan of the nested model outputs being handled that way, we should IMO have something like 
```python 
def seperate_model_outputs(model_outputs):
      if "encoder_outputs" in model_outputs:
             model_outputs["encoder_outputs"] = split(model_outputs["encoder_outputs"])
      model_outputs = split(model_outputs) # ignores the ModelOutput type that is already split
      return model_outputs
```
since the only model output you ever have is the `encoder_outputs` this to me makes it more readable and clearer! 
It adds an extra function but should make it easier to maintain 🤗